README.rst
===============================
Slipstream
===============================

.. TODO add these badges back, \o/
.. .. image:: https://badge.fury.io/py/draftin_a_flask.png
    :target: http://badge.fury.io/py/draftin_a_flask
    
.. .. image:: https://travis-ci.org/waynew/draftin-a-flask.png?branch=master
        :target: https://travis-ci.org/waynew/draftin-a-flask

.. .. image:: https://pypip.in/d/draftin_a_flask/badge.png
        :target: https://crate.io/packages/draftin_a_flask?version=latest

Slipstream is a cool tool that allows you to generate static blogs from
`Draft`_ `WebHooks`_

.. _Draft: http://draftin.com
.. _WebHooks: https://draftin.com/documents/69898?token=5fjKKlZ0-AeBzqj_RAftAGdzRzl9VBfBHj5wpSWm_gU


Quickstart (Easiest Version)
----------------------------

Run the following commands:

::

    $ docker run -d --name slipstream -p '0.0.0.0:5000:5000' \
      -e SLIPSTREAM_OUTPUT_DIR='/usr/share/nginx/html' waynew/slipstream
    $ docker run -d --name webserver --volumes-from slipstream \
      -p '0.0.0.0:8080:80' nginx
    $ docker logs --tail=1 slipstream
    Api Key: vsup5kabBC5Qen2ADH1NMdnGNdgkZ3bXlNUJFZDLdbg=
    $ curl icanhazip.com
    203.0.113.42

Copy the key and your IP address and head on over to the Draft `publishers`_
page, and click the ``WebHook URL`` link. Add your URL there, go to a draft
document and click the publish link. You should now see your blog posts when
you go http://localhost:8080.

.. _publishers: https://draftin.com/publishers

TODO:

Actually write some code. Also add more documentation for doing things like
changing up the templates.

.. .. _WebHooks: https://draftin.com/documents/69898?token=5fjKKlZ0-AeBzqj_RAftAGdzRzl9VBfBHj5wpSWm_gU)

.. * Free software: BSD license
.. * Documentation: http://draftin_a_flask.rtfd.org.

.. Usage
.. -----

.. ::

..     $ pip install draftin_a_flask
    $ env DIF_CONTENT=/path/to/content \
    DIF_OUTPUT=/path/to/output \
    DIF_PELICAN=/path/to/pelican_binary draftican
    Listening at endpoint QRFky1tR0KqHGM3cJoitwEi8tTpknaNnMpNHHiTIm8
    * Running on http://0.0.0.0:5678/
    * Restarting with reloader
    Listening at endpoint QRFky1tR0KqHGM3cJoitwEi8tTpknaNnMpNHHiTIm8
    
.. Setup your WebHook from within Draft, and now you can write your blog posts in
.. Draft and easily publish.


.. Future Features
.. ---------------

.. * Automatic uploads using rsync/ssh/file copy
.. * Settings provided in a file
.. * Improved error handling (e.g. missing title, etc.)


.. Known Bugs
.. ----------

.. * If you're missing important fields (like title and date) it probably will
..   skip publishing that doc.


HISTORY.rst
.. :changelog:

Release History
===============

dev (XXXX)
----------

**Major Features**

- Working on tests and the server itself


tests/test_util.py
import base64
import logging
import os
import pytest
import slipstream.util
import string
import struct
import tempfile
from contextlib import contextmanager
from hypothesis import given
from hypothesis.strategies import text, integers
from unittest import mock

log = logging.getLogger('slipstream')
log.addHandler(logging.StreamHandler())
log.setLevel(logging.DEBUG)

try:
    FileNotFoundError
except:
    FileNotFoundError = OSError


@contextmanager
def backup_file(filename):
    with tempfile.TemporaryFile(mode='w+') as bak:
        if not os.path.exists(filename):
            try:
                with open(filename, 'w+') as f:
                    yield f
            finally:
                try:
                    os.unlink(filename)
                except FileNotFoundError:
                    pass  # Whatever
        else:
            try:
                with open(filename, 'r+') as f:
                    bak.write(f.read())
                    f.seek(0)
                    yield f
            finally:
                bak.flush()
                bak.seek(0)
                with open(filename, 'w') as f:
                    f.write(bak.read())


@given(text(alphabet=string.printable, min_size=1))
def test_if_SLIPSTREAM_API_KEY_exists_it_should_be_used(monkeypatch, txt):
    expected_key = txt
    monkeypatch.setitem(os.environ, 'SLIPSTREAM_API_KEY', expected_key)
    
    actual_key = slipstream.util.get_api_key()
    assert actual_key == expected_key


def test_if_SLIPSTREAM_API_KEY_does_not_exist_and_SLIPSTREAM_API_KEYFILE_points_to_bad_file_it_should_FileNotFoundError(monkeypatch):
    try:
        monkeypatch.delitem(os.environ, 'SLIPSTREAM_API_KEY')
    except KeyError:
        pass  # Ok, ENV doesn't have that variable.

    # Make sure we have a file that doesn't exist
    with tempfile.NamedTemporaryFile(delete=True) as f:
        filename = f.name
    monkeypatch.setitem(os.environ, 'SLIPSTREAM_API_KEYFILE', filename)

    with pytest.raises(FileNotFoundError):
        slipstream.util.get_api_key()


@given(text(alphabet=''.join(string.printable.split()), min_size=1))
def test_if_SLIPSTREAM_API_KEY_does_not_exist_and_SLIPSTREAM_API_KEYFILE_points_to_good_file_it_should_use_that_key(monkeypatch, txt):
    expected_key = txt
    try:
        monkeypatch.delitem(os.environ, 'SLIPSTREAM_API_KEY')
    except KeyError:
        pass  # Ok, ENV doesn't have that variable.

    with tempfile.NamedTemporaryFile(mode='w', delete=True) as f:
        f.write(expected_key)
        f.flush()
        monkeypatch.setitem(os.environ, 'SLIPSTREAM_API_KEYFILE', f.name)

        actual_key = slipstream.util.get_api_key()

    assert actual_key == expected_key


@given(text(alphabet=''.join(string.printable.split()), min_size=1))
def test_if_no_env_vars_exist_but_api_key_exists_then_use_it(monkeypatch, txt):
    expected_key = txt
    for var in ('SLIPSTREAM_API_KEY', 'SLIPSTREAM_API_KEYFILE'):
        try:
            monkeypatch.delitem(os.environ, var)
        except KeyError:
            pass  # Ok, ENV doesn't have that variable.

    with backup_file('.slipstream_api_key') as f:
        print('writing {!r}'.format(expected_key))
        f.write(expected_key)
        f.truncate()
        f.close()

        actual_key = slipstream.util.get_api_key()

    assert actual_key == expected_key


@given(integers(max_value=9223372036854775))
def test_if_no_env_vars_exist_and_no_file_exists_it_should_be_created_and_urlsafe64encode_output_dumped_in(monkeypatch, num):
    print(num)
    expected_key = base64.urlsafe_b64encode(struct.pack('@l', num)).decode()
    for var in ('SLIPSTREAM_API_KEY', 'SLIPSTREAM_API_KEYFILE'):
        try:
            monkeypatch.delitem(os.environ, var)
        except KeyError:
            pass  # Ok, ENV doesn't have that variable.

    with mock.patch('slipstream.util.os.urandom') as fake_urandom:
        fake_urandom.return_value = struct.pack('@l', num)

        with backup_file('.slipstream_api_key') as f:
            os.unlink(f.name)

            actual_key = slipstream.util.get_api_key()
            key_in_file = open(f.name).read()

    assert actual_key == expected_key
    assert actual_key == key_in_file


tests/test_slipstream.py
import json
import logging
import os
import pytest
from unittest import mock

API_KEY = 'fnord'
os.environ['SLIPSTREAM_API_KEY'] = API_KEY

from slipstream import slipstream

log = logging.getLogger('slipstream')
log.addHandler(logging.StreamHandler())
log.setLevel(logging.DEBUG)


# TODO: Module? Or not? -W. Werner, 2015-11-21
@pytest.fixture(scope="module")
def client():
    return slipstream.app.test_client()


def test_client_should_use_api_key():
    assert slipstream.app.config['API_KEY'] == API_KEY


def test_post_should_fail_if_api_key_is_wrong(client):
    rv = client.post('/{}-invalid'.format(API_KEY))
    assert rv.status_code == 403


def test_post_should_pass_payload_to_core_publish(client):
    expected_id = 'Some document id thing'
    expected_title = 'The Ballad of Wol Emulov'
    expected_content = '## This is some markdown'
    expected_author = 'roscivs@indessed.com'
    payload = dict(
        id=expected_id,
        name=expected_title,
        content=expected_content,
        content_html='<html><h2>This is some markdown</h2></html>',
        user={'id': 42, 'email': expected_author},
        created_at='2010-08-14T09:23:12-05:00',
        updated_at='2010-08-14T09:23:12-05:00',
    )

    with mock.patch('slipstream.core.publish') as fake_publish:
        rv = client.post('/{}'.format(API_KEY), data={
            'payload': json.dumps(payload),
        })

        assert rv.status_code == 200
        fake_publish.assert_called_with(
            id=expected_id,
            title=expected_title,
            content=expected_content,
            author=expected_author,
        )


slipstream/util.py
import base64
import logging
import os

log = logging.getLogger(__name__)


def get_api_key():
    '''
    Get API key from environment, file system, or generate it if not found.

    Look first in the environment variable ``SLIPSTREAM_API_KEY``. If that is
    not found, then check for ``SLIPSTREAM_API_KEYFILE``.  If the
    ``SLIPSTREAM_API_KEYFILE`` *does* exist and it is not a file, raise
    ``FileNotFoundError`` on python3, ``IOError`` on python2.  If that variable
    does not exist then look for a file named `.slipstream_api_key` in the
    current directory.

    If that file does not exist, attempt to create it and insert a random
    32-bit url-safe base64 encoded string.
    '''

    key = os.environ.get('SLIPSTREAM_API_KEY')
    if key is None:
        log.info('No SLIPSTREAM_API_KEY')
        filename = os.environ.get('SLIPSTREAM_API_KEYFILE')
        if filename is not None:
            log.info('SLIPSTREAM_API_KEYFILE=%r', filename)
            with open(filename) as keyfile:
                key = keyfile.read()
        else:
            filename = '.slipstream_api_key'
            log.info('No SLIPSTREAM_API_KEY* vars found,'
                     ' trying %r', filename)
            with open(filename, 'a+') as keyfile:
                keyfile.seek(0)
                key = keyfile.read()
                log.debug('Found %r in %r', key, filename)
                if not key:
                    log.info('No key found in file, generatig key')
                    keyfile.write(
                        base64.urlsafe_b64encode(os.urandom(32)).decode()
                    )
                    keyfile.flush()
                    keyfile.seek(0)
                    key = keyfile.read()

    return key


slipstream/slipstream.py
'''
Copyright 2015 - Wayne Werner

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

import json
from flask import Flask, abort, request
from . import util
from . import core

app = Flask(__name__)
app.config['API_KEY'] = util.get_api_key()


@app.route('/<api_key>', methods=['POST'])
def api(api_key):
    if api_key != app.config['API_KEY']:
        app.logger.warning('Bad API key %r', api_key)
        abort(403)
    else:
        payload = json.loads(request.form['payload'])
        # TODO: Also pass created/updated dates -W. Werner, 2015-11-21
        core.publish(id=payload['id'],
                     title=payload['name'],
                     content=payload['content'],
                     author=payload['user']['email'],
                     )

        app.logger.debug(payload)
        return 'OK'
